crates/ra_ide/src/syntax_highlighting.rs

```diff
@@ -79,6 +89,15 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
 
     let mut in_macro_call = None;
 
+    // Determine the root based on the range
+    let root = match range {
+        Some(range) => match root.covering_element(range) {
+            NodeOrToken::Node(node) => node,
+            NodeOrToken::Token(token) => token.parent(),
+        },
+        None => root,
+    };
+
     for event in root.preorder_with_tokens() {
         match event {
```

Comment:
Let's also add a quick 

```
if element.text_range().intersection(range_to_highlight).is_none() { continue }
```

Reply:
Where would you like that? On the LSP side the server is allowed to return all highlights for the file if it can't do the range for whatever reason.

Reply:
For every event: if parent happens to be a big node, like an impl, we might process unnecessary children